### PR TITLE
Avoid adding then dropping FK constraints

### PR DESCRIPTION
**Pull request by greg0ire**
Previously, when processing entity relationships, SchemaTool would:

1. Add FK constraints as they were encountered
2. If a conflict was detected (same columns, different target), drop the previously added FK
3. Blacklist the composite key to prevent adding any FK for those columns

This add-then-drop pattern is not great, and will make things even worse when migrating to the TableEditor API, because it will require known the exact auto-generated FK name, which in turn would force copying the DBAL algorithm for generating FK names.

Solution:
Implement a two-pass approach:
1. Collection Phase: Gather all FK metadata without adding to tables
   - Detect conflicts between FKs (same local columns, different targets)
   - Mark conflicting composite keys as blacklisted
   - Store FK metadata for non-conflicting cases
2. Application Phase: Add only non-conflicting FKs to their tables
   - Skip any composite keys that were blacklisted due to conflicts

### CODE REVIEW

File: src/Tools/SchemaTool.php [hunk 1]
```diff
@@ -499,6 +501,43 @@ public function getSchemaFromMetadata(array $classes): Schema
...
+                            $tableEditor->addForeignKeyConstraint(new ForeignKeyConstraint(
+                                localColumnNames: $fkData['localColumns'],
+                                foreignTableName: $fkData['foreignTableName'],
+                                foreignColumnNames: $fkData['foreignColumns'],
+                                options: $fkData['fkOptions'],
+                            ));
+                        },
+                    )->create();
+                // @phpstan-ignore class.notFound (Using unreleased Schema::edit() API)
+                } catch (InvalidSchemaModification) {
```
Comment (greg0ire) [inline src/Tools/SchemaTool.php]: @morozov I think this might be one place where the "you know what you added" (or in this case, removed) breaks down: here the user might be the one that removed the table, thanks to the `postGenerateSchemaTable` event above.

Comment (morozov) [reply src/Tools/SchemaTool.php]: I think if it breaks down here, that points to an ORM flaw rather than a gap in the editor API design. We could add the needed methods to the editor, but in this specific case the ORM could call `$schema->hasTable()` and check whether the table is still there.

More generally, this feels like a smell similar to the one being fixed: the ORM defines a schema with a table, the user drops it, and then the ORM has to recalibrate and figure out what the user did. Maybe the FK constraints should be added to the effective/final schema, rather than to the draft before the user's edits?

One more note on catching `InvalidSchemaModification`: it's a logic exception — essentially "you're doing something wrong" — so it isn't meant to be caught and handled.

Comment (greg0ire) [reply src/Tools/SchemaTool.php]: Not sure how I missed that method :sweat: … thanks!

As for the flaw, the event I mention is fired after each table is added:

https://github.com/doctrine/orm/blob/883c76a229793adfdd379a2477e4ab268108889e/src/Tools/SchemaTool.php#L479-L483

 So maybe I'm breaking it right now by deferring the addition of the foreign key like I did: the event listener no longer has access to the foreign keys :see_no_evil: 

Maybe that's an acceptable breaking change, let us see what other maintainers think.

File: tests/Tests/ORM/Functional/SchemaTool/MySqlSchemaToolTest.php [hunk 2]
```diff
@@ -102,13 +102,48 @@ public function testGetCreateSchemaSql(): void
...
-        self::assertEquals('ALTER TABLE cms_users_tags ADD CONSTRAINT FK_93F5A1ADA76ED395 FOREIGN KEY (user_id) REFERENCES cms_users (id)', $sql[11]);
-        self::assertEquals('ALTER TABLE cms_users_tags ADD CONSTRAINT FK_93F5A1ADBAD26311 FOREIGN KEY (tag_id) REFERENCES cms_tags (id)', $sql[12]);
-        self::assertEquals('ALTER TABLE cms_addresses ADD CONSTRAINT FK_ACAC157BA76ED395 FOREIGN KEY (user_id) REFERENCES cms_users (id)', $sql[13]);
-        self::assertEquals('ALTER TABLE cms_phonenumbers ADD CONSTRAINT FK_F21F790FA76ED395 FOREIGN KEY (user_id) REFERENCES cms_users (id)', $sql[14]);
+        self::assertThat($sql[8], self::logicalOr(
+            // DBAL < 4.5
+            self::equalTo('ALTER TABLE cms_users ADD CONSTRAINT FK_3AF03EC5A832C1C9 FOREIGN KEY (email_id) REFERENCES cms_emails (id)'),
+            // DBAL 4.5
+            self::equalTo('ALTER TABLE cms_users ADD FOREIGN KEY (email_id) REFERENCES cms_emails (id)'),
+        ));
```
Comment (greg0ire) [inline tests/Tests/ORM/Functional/SchemaTool/MySqlSchemaToolTest.php:110]: > The DBAL will generate the name in this case.

@morozov it looks like the name is not generated  now: there is no name at all. Am I doing something wrong?

Comment (morozov) [reply tests/Tests/ORM/Functional/SchemaTool/MySqlSchemaToolTest.php:110]: No, this is valid SQL. MySQL will generate the name automatically. All databases do that except for SQLite. In SQLite, an unnamed constraint stays unnamed (which is also valid).

> The DBAL will generate the name in this case.

This statement in https://github.com/doctrine/orm/pull/12523#discussion_r3523387942 was about the name of the implicit index backing the FK constraint (it is auto-generated). The legacy/mutating API (`Table::addForeignKeyConstraint()`) automatically generates both the constraint and index names. Constraint name generation is redundant, so the new API (immutable) doesn't generate it.

This works for both FK and unique constraints (see https://github.com/doctrine/dbal/pull/6668).

Comment (greg0ire) [reply tests/Tests/ORM/Functional/SchemaTool/MySqlSchemaToolTest.php:110]: @morozov the other thing I had to change is tests/Tests/ORM/Functional/Ticket/GH11501Test.php : it seems that now the name is generated by MySQL rather than by DBAL, MySQL generates a name that is too long:

```
1) Doctrine\Tests\ORM\Functional\Ticket\GH11501Test::testDeleteOneToManyCollectionWithSingleTableInheritance
Doctrine\ORM\Tools\ToolsException: Schema-Tool failed with Error 'An exception occurred while executing a query: Identifier name 'one_to_many_single_table_inheritance_test_entities_parent_join_ibfk_1' is too long' while executing DDL: ALTER TABLE one_to_many_single_table_inheritance_test_entities_parent_join ADD FOREIGN KEY (test_entity_holder_id) REFERENCES GH11501TestEntityHolder (id)

/home/runner/work/orm/orm/src/Tools/ToolsException.php:18
/home/runner/work/orm/orm/src/Tools/SchemaTool.php:106
/home/runner/work/orm/orm/tests/Tests/OrmFunctionalTestCase.php:835
/home/runner/work/orm/orm/tests/Tests/ORM/Functional/Ticket/GH11501Test.php:19

Caused by
Doctrine\DBAL\Exception\DriverException: An exception occurred while executing a query: Identifier name 'one_to_many_single_table_inheritance_test_entities_parent_join_ibfk_1' is too long

/home/runner/work/orm/orm/vendor/doctrine/dbal/src/Driver/API/MySQL/ExceptionConverter.php:102
/home/runner/work/orm/orm/vendor/doctrine/dbal/src/Connection.php:1453
/home/runner/work/orm/orm/vendor/doctrine/dbal/src/Connection.php:1389
/home/runner/work/orm/orm/vendor/doctrine/dbal/src/Connection.php:908
/home/runner/work/orm/orm/src/Tools/SchemaTool.php:104
/home/runner/work/orm/orm/tests/Tests/OrmFunctionalTestCase.php:835
/home/runner/work/orm/orm/tests/Tests/ORM/Functional/Ticket/GH11501Test.php:19

Caused by
Doctrine\DBAL\Driver\Mysqli\Exception\ConnectionError: Identifier name 'one_to_many_single_table_inheritance_test_entities_parent_join_ibfk_1' is too long

/home/runner/work/orm/orm/vendor/doctrine/dbal/src/Driver/Mysqli/Exception/ConnectionError.php:24
/home/runner/work/orm/orm/vendor/doctrine/dbal/src/Driver/Mysqli/Connection.php:60
/home/runner/work/orm/orm/vendor/doctrine/dbal/src/Driver/Middleware/AbstractConnectionMiddleware.php:34
/home/runner/work/orm/orm/vendor/doctrine/dbal/src/Logging/Connection.php:46
/home/runner/work/orm/orm/vendor/doctrine/dbal/src/Connection.php:906
/home/runner/work/orm/orm/src/Tools/SchemaTool.php:104
/home/runner/work/orm/orm/tests/Tests/OrmFunctionalTestCase.php:835
/home/runner/work/orm/orm/tests/Tests/ORM/Functional/Ticket/GH11501Test.php:19

Caused by
mysqli_sql_exception: Identifier name 'one_to_many_single_table_inheritance_test_entities_parent_join_ibfk_1' is too long

/home/runner/work/orm/orm/vendor/doctrine/dbal/src/Driver/Mysqli/Connection.php:58
/home/runner/work/orm/orm/vendor/doctrine/dbal/src/Driver/Middleware/AbstractConnectionMiddleware.php:34
/home/runner/work/orm/orm/vendor/doctrine/dbal/src/Logging/Connection.php:46
/home/runner/work/orm/orm/vendor/doctrine/dbal/src/Connection.php:906
/home/runner/work/orm/orm/src/Tools/SchemaTool.php:104
/home/runner/work/orm/orm/tests/Tests/OrmFunctionalTestCase.php:835
/home/runner/work/orm/orm/tests/Tests/ORM/Functional/Ticket/GH11501Test.php:19
```

So… this might be a breaking change :grimacing:

Comment (morozov) [reply tests/Tests/ORM/Functional/SchemaTool/MySqlSchemaToolTest.php:110]: Yeah, it looks like a MySQL ~~bug~~ "feature" ([Bug #13942](https://bugs.mysql.com/bug.php?id=13942)) which the previous behavior masked:
```sql
CREATE TABLE aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
  (id INT NOT NULL, parent_id INT NULL, PRIMARY KEY (id));

ALTER TABLE aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
  ADD FOREIGN KEY (parent_id) REFERENCES aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa (id);
-- ERROR 1059 (42000): Identifier name '..._ibfk_1' is too long
```

It also highlights the existing DBAL 4.x API inconsistency: it will auto-generate the name if the FK is added to the table via mutation (`Table::addForeignKeyConstraint()` accepts parameters, generates the name and constructs the FK) but not if it's passed to the constructor (accepts a pre-constructed object and does not generate the name, even if it's empty, so prone to the same issue).

The mutation-to-editor API migration is a breaking change and is opt-in from the DBAL 4.x perspective. ORM can either accept this as an edge-case break or work around it (generate the constraint name).

UPD: it's a MySQL-only issue (MariaDB still allows creation but truncates during introspection).

Comment (greg0ire) [reply tests/Tests/ORM/Functional/SchemaTool/MySqlSchemaToolTest.php:110]: I'd say it's an acceptable edge case, and that people running into this should upgrade to MySQL 8 or rename their tables before they attempt upgrading to ORM 3.7.

File: src/Tools/SchemaTool.php (hunk 1 above)
Comment (stof) [reply src/Tools/SchemaTool.php]: I never used the `postGenerateSchemaTable` event in my own projects. I always put the logic in `postGenerateSchema`

Comment (greg0ire) [reply src/Tools/SchemaTool.php]: Thanks for the insight. Also, inspecting foreign keys there seems like a far fetched scenario.

**Review by stof — APPROVED** [review]
